src/stream/src/executor/dispatch.rs

```diff
@@ -389,37 +398,85 @@ impl StreamConsumer for DispatchExecutor {
     type BarrierStream = impl Stream<Item = StreamResult<Barrier>> + Send;
 
     fn execute(mut self: Box<Self>) -> Self::BarrierStream {
+        let barrier_batch_size = self
+            .inner
+            .context
+            .config
+            .developer
+            .max_barrier_batch_size
+            .saturating_sub(1);
         #[try_stream]
         async move {
-            let input = self.input.execute();
-
-            #[for_await]
-            for msg in input {
+            let mut input = self.input.execute().peekable();
+            let mut end_of_stream = false;
+            while !end_of_stream {
+                let Some(msg) = input.next().await else {
+                    end_of_stream = true;
+                    continue;
+                };
+                let mut barrier_batch = vec![];
```

Comment:
Is it possible to decouple the logic for batching messages from dispatching them?